add ConcurrentOnHeapHnswGraph and Builder

jbellis opened the pull request:
### Motivation

1. It is faster to query a single on-heap hnsw index, than to query multiple such indexes and combine the result.
2. Even with some contention necessarily occurring during building of the index, we still come out way ahead in terms of total efficiency vs creating per-thread indexes and combining them, since combining such indexes boils down to "pick the largest and then add all the other nodes normally," you don't really benefit from having computed the others previously.

### Performance

On my i9-12900 (8 performance cores and 8 efficiency) I get a bit less than 10x speedup of wall clock time for building and querying the "siftsmall" and "sift" datasets from http://corpus-texmex.irisa.fr/.  The small dataset is 10k vectors while the large is 1M.  This speedup feels pretty good for a data structure that isn't completely parallelizable, and it's good to see that it's consistent as the dataset gets larger. 

The concurrent classes achieve identical recall compared to the non-concurrent versions within my ability to test it, and are drop-in replacements for OnHeapHnswGraph and HnswGraphBuilder; I use threadlocals to work around the places where the existing API assumes no concurrency.

The timing harness (which includes recall checking) is here: https://github.com/jbellis/hnswdemo/tree/concurrent

### Design

ConcurrentOnHeapHnswGraph is basically the same design as OnHeapHnswGraph but with the neighbor lists changed to use ConcurrentHashMap at both level 0 and higher levels.  The internal guarantees change a bit along with this, e.g., addNode only initializes a neighbor list for node N when N is added, rather than all nodes <= N.

ConcurrentHnswGraphBuilder changes more, because it’s designed around “all the logic is in Builder and it calls primitive operations on the Graph and the Neighbors as necessary.”  Unfortunately you can’t build threadsafe concurrent structures this way, not without massive synchronized blocks and a ton of contention.  So I’ve moved a lot of the logic around adding neighbors into ConcurrentNeighborSet.

The key to achieving good concurrency without synchronization is, we track in-progress node additions in a ConcurrentSkipListSet. After adding ourselves, we take a snapshot of this set, and consider all other in-progress updates as neighbor candidates (subject to normal level constraints).

Commit:
Use HashMap for OnHeapHnswGraph neighbors

Switch OnHeapHnswGraph from representing a single node's neighbors
as a TreeMap, to a HashMap, and update its callers to no longer assume that
NodeIterator results are ordered by ordinal.

Commit:
run tidy

Commit:
add pretty-printing to graph equality tests

Commit:
add comment

Commit:
add ConcurrentOnHeapHnswGraph and Builder

Tests run against the new Concurrent classes, except where that doesn't make sense

Commit:
using IntStream.iterate to sort doesn't work here, use a while loop and Collections.sort instead

jbellis commented:
TODO: get ram usage working

jbellis commented:
TODO: update test suite to test both Concurrent and normal classes.

Commit:
move prettyPrint to HnswGraphTestCase

Commit:
replace wildcard import with individual classes

msokolov commented:
Very exciting! 

After a very quick glance through the code, a few high-level comments/questions: 

To use this when building indexes in Lucene, I think we'd need a way to control the threads used, so e.g. I don't think the parallel streams would allow for that. Have you considered accepting an Executor and using that to run the updates? 

TBH I am not an expert on Lucene's IndexWriter threading model, but here is my stab at some background info: We need to maintain free threads that enable indexing to go on concurrent with flush / merge that would invoke this.  ConcurrentMergeScheduler, for one, enables callers to specify the number of threads used for merging. Typically these are allocated per-merged-segment, but in theory it can be free to allow a single merge to use multiple threads. As for flush (initial creation of a new segment), I believe this is always single-threaded today. Perhaps we could use concurrency there too somehow, but it's a secondary target I think.

I don't see the need for these classes to conform to the existing HnswGraph API. If we decide we want to use this implementation when writing the index, we can switch LucenexxHnswWriter to do so. Given that, could we drop the use of ThreadLocal?

I wonder what the single-threaded performance of these new classes is as compared to the existing impl. Have you tested that?

Commit:
simplify to a single Map of graph levels since we're not special-casing L0 anymore

Commit:
optimize CNS.nodeIterator

Commit:
un-rename eps (makes it easier to compare to non-concurrent class)

Commit:
avoid using stream() since it creates unnecessary Map.Entry objects

Commit:
add explicit variable types

Commit:
merge from main

Commit:
fix NPE in test sort

Commit:
rerun tidy

jbellis commented:
I've made a couple optimizations and merged to main.

With the optimizations it's still about 38% slower than the non-concurrent code (including the HashMap optimization for the non-concurrent that went in).  Nothing stands out in the profiler as a problem anymore; I guess that's what you get replacing HashMap with ConcurrentHashMap and more especially replacing ArrayList with ConcurrentHashMap in L0.

I still do not know how to fix the ram usage estimates; any suggestions?

Commit:
add ThreadPoolExecutor support

Commit:
replace do/while compareAndSet loops with getAndAccumulate

jbellis commented:
Replaced parallel stream with ExecutorService.  Only ThreadPoolExecutor is supported, since it knows how many thread it has available.  After a quick look that is the only type I see Lucene creating.

I was pleasantly surprised that the TPE is no slower than parallel streams in my testing.

tang-hi commented:
> I've made a couple optimizations and merged to main.
> 
> With the optimizations it's still about 38% slower than the non-concurrent code (including the HashMap optimization for the non-concurrent that went in). Nothing stands out in the profiler as a problem anymore; I guess that's what you get replacing HashMap with ConcurrentHashMap and more especially replacing ArrayList with ConcurrentHashMap in L0.
> 
> I still do not know how to fix the ram usage estimates; any suggestions?

I believe the issue is related to the Java Platform Module System, which was introduced in Java 9. You can modify the following section in the [gradlew script](https://github.com/apache/lucene/blob/3c163745bb07aed51b52878de0666f1405696147/gradlew#L221) to avoid the exception:
https://github.com/apache/lucene/blob/3c163745bb07aed51b52878de0666f1405696147/gradlew#L221

**modified version**
```shell
JVM_ARGS=" -Ptests.jvmargs=\"--add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.util.concurrent.atomic=ALL-UNNAMED\""

# Collect all arguments for the java command, following the shell quoting and substitution rules

eval set -- $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS" "$JVM_ARGS"
```
After making this change, the exception should no longer be thrown. However, the test may still fail due to an AssertionError, as shown below:

```
  java.lang.AssertionError: expected:<1185336.0> but was:<486483.0>
   >         at __randomizedtesting.SeedInfo.seed([A41F62241729FCF:CF0592627AF94878]:0)
   >         at org.junit.Assert.fail(Assert.java:89)
   >         at org.junit.Assert.failNotEquals(Assert.java:835)
   >         at org.junit.Assert.assertEquals(Assert.java:555)
   >         at org.junit.Assert.assertEquals(Assert.java:685)
   >         at org.apache.lucene.util.hnsw.HnswGraphTestCase.testRamUsageEstimate(HnswGraphTestCase.java:789)
   >         at org.apache.lucene.util.hnsw.TestHnswFloatVectorGraph.testRamUsageEstimate(TestHnswFloatVectorGraph.java:37)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >         at java.base/jdk.internal.reflect.NativeMethodAccesso
```

Commit:
add JVM_ARGS to open jdk classes for use by ram usage estimator

jbellis commented:
Thanks! For me it is working:

```
$ ./gradlew -p lucene/core test --tests "*Hnsw*"
...
:lucene:core:test (SUCCESS): 75 test(s)
```

Note that adding these args do not seem to be picked up by intellij when running tests, is there a better place to add them?

Commit:
use vectorsCopy in scoreBetween

Commit:
add missing license headers

Commit:
make linter happy

Commit:
fix ram estimation

Commit:
also update ram estimation for OHHG after hashmap change

Commit:
Revert "add JVM_ARGS to open jdk classes for use by ram usage estimator"

This reverts commit 1b08a749298f0559f4ead3902e871728ef8e4393.

Commit:
open jdk classes in a way that intellij can use

Commit:
Unlike classic OnHeapHnswGraph, our ram usage depends on how many neighbors we have (we don't pre-allocate a set large enough for maximum neighbor count because the JDK exposes no such constructors)

jbellis commented:
thanks again.

and you're right, the ram usage needed a rewrite, it is passing now with the seed you found.

Commit:
add factory classes to allow running concurrent and serial hnsw graph builder tests from the same code

MarcusSorealheis reviewed (commented):
Sweet. I'm excited to test and possibly contribute to `ConcurrentOnHeapHnswGraph` when Java21 is released. There may be some observability opportunity beyond the info stream, which is great today.

jbellis commented on lucene/core/src/java/org/apache/lucene/util/hnsw/OnHeapHnswGraphBuilder.java:
```diff
@@ -0,0 +1,431 @@
+    double randDouble;
+    do {
+      randDouble = random.nextDouble(); // avoid 0 value, as log(0) is undefined
+    } while (randDouble == 0.0);
```
This is part of the original, non-concurrent code -- it shows up in the diff because I renamed it to extract HnswGraphBuilder as an interface.  So I'd prefer to leave tweaking it out of scope for this PR as much as possible.

jbellis commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentOnHeapHnswGraph.java:
```diff
@@ -0,0 +1,280 @@
+      // we expect there to be nodesOnLevel / levelLoadFactor Nodes in its internal table.
+      // there is also an entrySet reference, 3 ints, and a float for internal use.
+      int nodeCount = (int) (numNodesOnLevel / chmLoadFactor);
+      long chmSize =
```
TBH I don't think it improves things much since most of the awkwardness comes from RamUsageEstimator.NUM_BYTES_OBJECT_REF being such a long string that you can't fit any other terms with it on the same line.

jbellis replied:
but I did add local aliases to those constants for readability

Commit:
add local aliases to constants for readability

jbellis commented:
how do we get this unblocked?  I don't have write access to request a reviewer.

Commit:
one more tidy

Commit:
fix errorProne crashing the build for warnings that were there before I started

Commit:
add missing @Override

MarcusSorealheis commented:
@jbellis while I am not the gate keeper to this kingdom, my only take would say patience would be the number one factor. It's a big change. My guess is that Patrick, Sokolov and many others are watching. Thanks for the adding the aliases.

msokolov commented:
Yep, I plan to review, definitely interested to see this get committed, but it's a bit complex and I need to find some quiet time, which is rare. You know we're all volunteers here!

Commit:
keep the original name of HnswGraphBuilder; add IHnswGraphBuilder for the interface (sorry)

Commit:
tidy

jbellis commented:
Thanks, Michael!

msokolov commented:
uh-oh this PR metastasized a bit iin the past few days. I think it was a better plan to commit the graph builder util class on its own prior to cutting over the codec usage. Can we separate these into two stages? I expect a lot of discussion around the codec usage/index writer integration and I don't want it to block getting the algorithmic change in.

On the other hand, since you have done the integration, are you able to test it using `KnnGraphTester` (recently moved to `luceneutil` package)? Or with some other tool that tests the whole integration, like `ann-benchmarks`?

msokolov commented:
uh-oh this PR metastasized a bit iin the past few days. I think it was a better plan to commit the graph builder util class on its own prior to cutting over the codec usage. Can we separate these into two stages? I expect a lot of discussion around the codec usage/index writer integration and I don't want it to block getting the algorithmic change in.

On the other hand, since you have done the integration, are you able to test it using `KnnGraphTester` (recently moved to `luceneutil` package)? Or with some other tool that tests the whole integration, like `ann-benchmarks`?

msokolov commented:
uh sorry, I got confused by the HnswSearcher->OnHeapHnswSearcher renaming

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+    Semaphore semaphore = new Semaphore(pool.getMaximumPoolSize());
+
+    for (int i = 0; i < vectorsToAdd.size(); i++) {
+      final int node = i; // copy for closure
```
I see - here is where we guarantee that each node is added once. They might be added out of sequence now, but that's OK

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+    // normalization factor for level generation; currently not configurable
+    this.ml = M == 1 ? 1 : 1 / Math.log(1.0 * M);
+    this.hnsw = new ConcurrentOnHeapHnswGraph(M);
+    this.graphSearcher =
```
Instead of relying on ThreadLocal can we manage this thread -> searcher association explicitly? EG in addVectors can we create a collection of per-thread builder objects (like the views we have on the graph) and then manage them ourselves? could even be a map of threadid -> builder-view. In this project we've had issues with GC not capturing ThreadLocal storage, eg see https://markmail.org/thread/fbklvo4tkdd5d5u5. If the storage is managed locally as a reference from the builder, it should get cleaned up properly.

jbellis replied:
No problem, I will remove the threadlocals.

jbellis replied:
Can do.

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+              try {
+                addGraphNode(node, vectorsToAdd);
+              } catch (IOException e) {
+                throw new RuntimeException(e);
```
We need to find a way to propagate these exceptions to the caller so we can re-throw. Also, UncheckedIOException is usually preferable for wrapping IOExceptions.

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+              }
+            });
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
```
Lucene generally uses IOException for most low-level failures and doesn't have any mechanism for handling uncaught unchecked exceptions, so we should convert this to an IOException and add throws to all the methods here that need it.

jbellis replied:
JDK documentation says that InterruptedException is only thrown when someone calls Thread.interrupt(), and we never do that.  Okay to just make it IllegalStateException?

jbellis replied:
... it looks like Lucene prefers to use its own ThreadInterruptedException

jbellis replied:
Got it.

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+
+    // Wait for any remaining tasks to complete
+    pool.shutdown();
+    try {
```
Perhaps we could return a Future and let the caller handle the details of how long to wait and managing the ThreadExecutor that *they supplied to us*. We don't want to be closing the pool here.

jbellis replied:
Good point, thanks for catching that.

jbellis replied:
I like the idea of returning a Future.

msokolov reviewed (commented):
I added some comments but I haven't nearly had a chance to review everything - it's a lot ! So I think we can make more rapid progress if we scale this back to the scope it had earlier - just the new concurrent graph builder - without attempting to integrate with the existing API and tests. Maybe initially we copy-paste the tests to make concurrent versions and then later if possible we can refactor?

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+  // we need two sources of vectors in order to perform diversity check comparisons without
+  // colliding
+  private final RandomAccessVectorValues<T> vectorsCopy;
+  private final AtomicBitSet initializedNodes;
```
calling this `initializedNodes` confused me since the usage here is really different from the variable of the same name in `HnswGraphBuilder` that is populated when initializing from another graph, but here we add to this set while adding nodes in `addNode`. I'm actually unclear why we need this and the check in `addGraphNode`. Aren't we guaranteed to insert each node exactly once?

jbellis replied:
Ahhhhhh that makes sense, it wasn't clear to me that under normal condition's it's the caller's responsibility to not add the same node multiple times.

jbellis commented on lucene/core/src/java/org/apache/lucene/util/hnsw/ConcurrentHnswGraphBuilder.java:
```diff
@@ -0,0 +1,329 @@
+    if (infoStream.isEnabled(HNSW_COMPONENT)) {
+      infoStream.message(HNSW_COMPONENT, "build graph from " + vectorsToAdd.size() + " vectors");
+    }
+    if (!(pool instanceof ThreadPoolExecutor)) {
```
I could do that, but the rest of Lucene always writes TPE construction as

      ExecutorService executorService =
          Executors.newFixedThreadPool(threadCount, new NamedThreadFactory("async-check-index"))
          
(because newFixedThreadPool returns ES not TPE) so it seemed cleaner to me to push the cast into my code  instead of making all callers do it.  Is that okay?

Commit:
explicilty track futures in flight instead of relying on pool.awaitTermination

Commit:
throw ThreadInterruptedException instead of RuntimeException

Commit:
capture thread and lambda exceptions instead of re-throwing RuntimeException

Commit:
remove ThreadLocal

Commit:
unnecessary to 'defend' against same node being added multiple times

Commit:
remove new interfaces and factory classes

Commit:
copy tests for concurrent classes

Commit:
use explicit getView instead of threadlocal View

Commit:
r/m create() overload that matches non-concurrent Builder

Commit:
tidy

Commit:
tweak to use accumulateAndGet instead of compareAndSet

jbellis commented:
Updated.

Commit:
merge from main

Commit:
visited set needs to be able to grow if nodes are added during search

Commit:
fix AtomicBitSet for concurrent calls to expandStorage

Commit:
move ThrowingBiFunction and ThrowingBiConsumer to ConcurrentNeighborSet

Commit:
add failing test

Commit:
make test actually fail by allowing jvm to use multiple cores

Commit:
fix race condition when two nodes add each other as neighbors concurrently

jbellis reviewed (commented):
What is a better way to allow the concurrent tests to use multiple cores?

Commit:
assert that we don't add a node as its own neighbor

Commit:
new test case that fails more reliably when concurrent graph ends up in a broken state

Commit:
propagate all throwables from executor, not just IOException

Commit:
validate that entry node is sane after construction

Commit:
Fix concurrency issues:

- Don't read entry node and level separately
- Link new nodes into graph bottom-up so that they don't affect other searches following the index down
- Compute diversity for fully-added and in-progress-being-added nodes separately
- Only compare diversity with other nodes from the same "group"

More details in comments.

jbellis commented:
It turns out that allowing test JVMs to use multiple cores uncovered more bugs.  (The test failure above happens about 2/3 of the time.)  I think I have fixes lined up (and tests that break more reliably) but it is getting late, I will review tomorrow and push when my head is clear.

Commit:
tidy

Commit:
snapshot isolation

    // It is tempting, but incorrect, to try to provide "adequate" isolation by
    // (1) keeping a bitset of complete nodes and giving that to the searcher as nodes to
    // accept -- but we need to keep incomplete nodes out of the search path entirely,
    // not just out of the result set, or
    // (2) keeping a bitset of complete nodes and restricting the View to those nodes
    // -- but we needs to consider neighbor diversity separately for concurrent
    // inserts and completed nodes; this allows us to keep the former out of the latter,
    // but not the latter out of the former (when a node completes while we are working,
    // that was in-progress when we started.)
    // The only really foolproof solution is to implement snapshot isolation as
    // we have done here.

Commit:
instead of assuming ThreadPoolExecutor, explicitly specify concurrentTasks as an argument to buildAsync

this also fixes the build(vectors, false) overload

Commit:
remove fully-connected test (turns out that it not actually expected for HNSW graphs even at fairly modest sizes) and changes to SelectiveOrds test (past ~200 vectors randomness starts causing failures)

Commit:
clean up TODO

Commit:
add logical clock size to ramBytesUsed

Commit:
add logical clock size to ramBytesUsed

Commit:
now that we have snapshot isolation we can switch back to building bottom-down

Commit:
we don't actually need thread safety in the Searcher bitset, just the ability to grow itself

extract JavaUtilBitSet -> GrowableBitSet to handle this use case

this makes concurrent build about 8% faster (~162s on Texmex benchmark to ~150)

Commit:
AtomicBitSet is no longer needed

Commit:
Merge branch 'concurrent3' of github.com:jbellis/lucene into concurrent3

Commit:
use a NeighborArray in ConcurrentNeighborSet instead of ConcurrentSkipListSet

CSLS handles concurrent inserts better, but reads are much more expensive:
- Iterate references instead of array
- Unbox from Long
- decode to int

Even in a construction-only workload, there are so many more reads than updates that it's better to optimize for reads

Commit:
optimize GrowableBitSet.clear for the case when we're clearing the entire set

Commit:
tidy

Commit:
fix NPE in ramByteUsed when graph levels have not been completely filled in

Commit:
fix merge

Commit:
fix NPE when size is called on empty graph

Commit:
create CompletionTracker to avoid box/unbox overhead of ConcurrentHashMap<Integer, Integer>

Commit:
add BitSet.clear() (#12268)

Commit:
improve comments

Commit:
the only time we would read an undefined node id is when we're inserting at the end, so we can just check size instead of the performance hit of filling unused slots with -1

Commit:
refactor ramBytesUsed to allow us to compute the incremental ram used by a new node

Commit:
tidy

Commit:
add searchConcurrent method that uses a GrowableBitSet since the graph may get added to as we search

Commit:
CompletionTracker had a race condition where ensureCapacity could replace the array reference while other threads updated the old one.  Fix using StampedLock

Commit:
add custom functional interface to avoid boxing in scoreBetween

Commit:
reformat

Commit:
re-use candidates NeighborQueue across calls to addGraphNode

Commit:
add insertsInProgress

Commit:
avoid the temptation to optimize L0 by using ArrayNodesIterator

Commit:
use a bitset in insertDiverse instead of allocating a new neighborarray

jbellis commented:
Closing in favor of #12421.
